File: coding/src/com/algos/core/li09_linkedlist/revised/L160_IntersectionOfTwoLinkedLists.java
package  com.algos.core.li09_linkedlist.revised;

import com.algos.core.models.ListNode;

public class L160_IntersectionOfTwoLinkedLists {
    public static void main(String[] args) {

    }

    // Problem: https://leetcode.com/problems/intersection-of-two-linked-lists/
    // Idea: Exchange curr1 and curr2 when they reach end of their respective lists.
    // They will meet at the intersection node.
    public ListNode getIntersectionNode(ListNode head1, ListNode head2) {
        ListNode curr1 = head1, curr2 = head2;
        while(curr1 != curr2){
            if(curr1 != null) {
                curr1 = curr1.next;
            } else{
                curr1 = head2;
            }

            if(curr2 != null){
                curr2 = curr2.next;
            } else{
                curr2 = head1;
            }
        }
        return curr1;
    }
}

File: dsa/patterns/09_linkedlist/python/l160_intersection_of_two_linked_lists.py
"""
Problem: https://leetcode.com/problems/intersection-of-two-linked-lists/

Idea:
Exchange curr1 and curr2 when they reach end of their respective lists.
They will meet at the intersection node.
"""

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def get_intersection_node(head1: ListNode, head2: ListNode) -> ListNode:
    curr1, curr2 = head1, head2

    while curr1 is not curr2:
        curr1 = curr1.next if curr1 else head2
        curr2 = curr2.next if curr2 else head1

    return curr1


# Example usage
if __name__ == "__main__":
    # Create intersecting lists
    # A: 1 -> 2 -> 3 \
    #                   -> 8 -> 9
    # B:       4 -> 5 /
    intersect = ListNode(8, ListNode(9))

    a1 = ListNode(1, ListNode(2, ListNode(3, intersect)))
    b1 = ListNode(4, ListNode(5, intersect))

    node = get_intersection_node(a1, b1)
    print(node.val if node else None)  # 8
